File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length = getLength(head);
        ListNode* dummy = new ListNode(-1);
        ListNode* ptr = dummy;
        ptr->next = head;
        int k = length - n;
        
        while(k-- > 0){
            ptr = ptr->next;
        }
        
        ptr->next = ptr->next->next;
        
        return dummy->next;
    }
    
private:
    int getLength(ListNode* head){
        int length = 0;
        ListNode* ptr = head;
        
        while(ptr != nullptr){
            length++;
            ptr = ptr->next;
        }
        
        return length;
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(n==1 && head.next==null){
            return null;
        }
       
        ListNode temp = new ListNode(-1);
        temp.next = head;
        ListNode slow = temp;
        ListNode fast = temp;
        
        while(n-- > 0){
            fast = fast.next;
        }
        
        while(fast!=null && fast.next!=null){
            fast = fast.next;
            slow = slow.next;
        }
        
        slow.next = slow.next.next;
        
        return temp.next;
    }
}

/*
ListNode start = new ListNode(0);
    ListNode slow = start, fast = start;
    slow.next = head;
    
    //Move fast in front so that the gap between slow and fast becomes n
    for(int i=1; i<=n+1; i++)   {
        fast = fast.next;
    }
    //Move fast to the end, maintaining the gap
    while(fast != null) {
        slow = slow.next;
        fast = fast.next;
    }
    //Skip the desired node
    slow.next = slow.next.next;
    return start.next;
*/